### PR TITLE
Add non-repudiation API

### LINKED ISSUE #8658: Non-repudiation: add a persistent key store with an upload interface
The current non-repudiation prototype has a volatile key store. Add a persistent key store and expose a way to upload new keys from remote processes (some kind of interface over HTTP should be sufficient as a first step).

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
The Web API for the non-repudiation proxy allows to retrieve the signed
payload for a given command, add a new certificate or query for an
existing one.

Closes #8658

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
Note: the tests currently only test the happy path, more tests for error conditions will be added before moving this out of draft (along with rebasing after #8981 is merged).

File: runtime-components/non-repudiation-api/src/test/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpointSpec.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        algorithm = AlgorithmString.wrap(algorithm),
+        fingerprint = FingerprintBytes.wrap(fingerprint),
+        payload = PayloadBytes.wrap(payload),
+        signature = SignatureBytes.wrap(signature),
+        timestamp = Instant.ofEpochMilli(timestamp),
+      )
+
+    SignedPayloadsEndpoint.toResponse(signedPayload) shouldBe SignedPayloadsEndpoint.Response(
+      algorithm = algorithm,
+      fingerprint = "ZmluZ2VycHJpbnQ=",
```
Comment (cocreature) [inline runtime-components/non-repudiation-api/src/test/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpointSpec.scala:41]: I assume that’s base64 encoded? maybe worth adding a comment to make the test a bit clearer (or just do the base64 encoding but then this test is pretty close to just replicating the logic which is also not super useful).

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/NonRepudiationApi.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      signedPayloadRepository,
+    ).map(_ => ())
+
+  final case class Configuration(
+      interface: String,
+      port: Int,
+      shutdownTimeout: FiniteDuration,
+  )
+
+  // Non-repudiation -> NR -> N = 78 and R = 82 in ASCII
```
Comment (cocreature) [inline runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/NonRepudiationApi.scala:40]: omg, is that how all our weird port numbers are constructed?

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import spray.json.{DefaultJsonProtocol, RootJsonFormat}
+
+import scala.util.Try
+
+private[api] final class CertificatesEndpoint private (certificates: CertificateRepository)
+    extends CertificatesEndpoint.JsonProtocol {
+
+  import CertificatesEndpoint._
+
+  private def putCertificate(certificate: X509Certificate): Route =
```
Comment (cocreature) [inline runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:24]: what happens if I add the same certificate twice?

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+  private def getCertificate(fingerprint: FingerprintBytes): Route =
+    handleExceptions(logAndReport(logger)(UnableToRetrieveTheCertificate)) {
+      certificates.get(fingerprint).fold(reject)(certificate => complete(encode(certificate)))
+    }
+
+  private val route: Route =
+    concat(
+      put {
```
Comment (cocreature) [inline runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:37]: Related question to the one above, if we’re not idempotent, we probably want post. Couldn’t find a test to check whether we are.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+object CertificatesEndpoint {
+
+  def apply(certificates: CertificateRepository): Route =
+    new CertificatesEndpoint(certificates).route
+
+  private val logger: Logger = LoggerFactory.getLogger(classOf[CertificatesEndpoint])
+
+  private[api] val InvalidCertificateString: String =
+    "Invalid upload, the 'certificate' field must contain a URL-safe base64-encoded X.509 certificate"
```
Comment (cocreature) [inline runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: what does url-safe mean here? I thought this is in the post body not a query param.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      algorithm = signedPayload.algorithm,
+      fingerprint = base64Url(signedPayload.fingerprint),
+      payload = base64Url(signedPayload.payload),
+      signature = base64Url(signedPayload.signature),
+      timestamp = signedPayload.timestamp.toEpochMilli,
+    )
+
+  private[api] trait JsonProtocol extends SprayJsonSupport with DefaultJsonProtocol {
+
+    protected implicit val apiSignedPayloadFormat: RootJsonFormat[Response] =
```
Comment (cocreature) [inline runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala:70]: for the JSON API and the trigger service we follow a common format of `result` and `error` fields and a `status` field in all responses. Have you considered adapting that here as well? If not, what do you think of doing so? totally fine not to do this at all or do it in a separate PR.

**Review by cocreature — APPROVED** [review]
Nice, thanks!

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/NonRepudiationApi.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/NonRepudiationApi.scala:40]: Wait. I assumed you knew about it.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 5 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: https://tools.ietf.org/html/rfc4648#page-7

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: I assumed this was common knowledge but I can add a link to the RFC if you think that makes sense.

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: (the assumption was towards the user, not towards you 😉)

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala (hunk 6 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala:70]: Sure, I was not aware of this. What is the `status` field though?

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:24]: Good question. My intention was to be idempotent but right know it's likely to fail on the backend, I'll add a test for this and fix it.

File: runtime-components/non-repudiation-api/src/test/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpointSpec.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/test/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpointSpec.scala:41]: Base64, URL-safe, I'll add a comment, thanks for pointing this out.

I preferred to not replicate the logic but rather have a "golden sample" of some sort. I was mostly interesting in making sure that this was not utterly broken.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 5 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: > I thought this is in the post body not a query param.

It would be `/v1/certificate/<fingerprint>`.

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: Whoops, sorry, I misread it. Yes, it's in the body. I just preferred to use a single encoding, does that make sense?

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala (hunk 6 above)
Comment (cocreature) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala:70]: It’s the status code. It makes more sense in the JSON API since annoyingly it can be different from the http status code in some cases (we stream some responses and we don’t know the status code when sending the header, fun side note: that’s why grpc uses http trailers instead of headers afaik).

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 5 above)
Comment (cocreature) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: Sorry for being unclear. I wasn’t asking what url-safe means, I was asking why we care about it if it’s not in the url and therefore also why we base64 encode it.

I’m not opposed to using base64 encoding here as well but I think you can make a reasonable argument that the ux is slightly nicer for sending the unencoded contents in something like a multi-part form request, e.g., you can pass it via `-F "parametername=@filename"` to curl whereas you have to call `base64` yourself.

How about the following approach for now:
1. Remove the comment on url-safe since it is misleading given that it is not in a url (the rest of the comment can stay)
2. Stick with base64 for now. We’re early access so we could break it if we want to and even if we don’t and we decide the multi-part is sufficiently nicer, keeping both around is relatively cheap.

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: Ok.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala (hunk 6 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala:70]: Ok.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 5 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: Mh, that would mean, however, that you cannot be guaranteed to be able to lookup a certificate (`GET /v1/certificate/<fingerprint>`) using the fingerprint returned when uploading (`PUT /v1/certificate`). Are you sure this is fine?

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: After some thinking, I'm not 100% sure about this. It would mean we would have places where we use base64 encoding and other where we use base64 URL-safe encoding (`GET /v1/certificate/<fingerprint>`). Are you sure it makes sense to use two different encodings? In particular, if we decide to use base64 encoding in all cases where a URL is not involved, it would mean we would return a base64 encoded fingerprint when uploading a certificate but then we would require a base64 URL-safe encoding to retrieve it.

Comment (cocreature) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: I think I’m confused, I thought this was only about the format used when uploading the certificate. I fully agree that the fingerprint returned in the response to that should be the same format as the one in the URL.

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: Sure, I was wondering if it makes sense to use two different encodings, even though it's not strictly needed to be URL-safe for certificates. Since we're forced to use URL-safe encoding for fingerprints, does it make sense to use a different encoding for certificates? I'm mostly trying to understand what an API user would think would make more sense.

Comment (cocreature) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: I guess I view the certificate as binary data which I already have (the certificate lies around somewhere locally) while I view the fingerprint as an opaque string that is returned and produced by the API. So I think it’s not entirely unreasonable to have different encodings for them. That said, I think two different base64 encodings are confusing as you mentioned and doesn’t solve anything So I think we have the following two options:

1. Use url-safe base64 encoding everywhere, i.e., do what you have. Probably just needs a comment somewhere on why we do this for cases where something is not in a url.
2. Use url-safe base64 encoding for things that are strings and/or used in urls, e.g., fingerprints. Upload binary data like certificates without base64 encoding them first, e.g., as multipart form requests.

I would suggest to go with 1 for now and keep the option open to later add/switch to 2 but I don’t feel strongly about this so if you disagree, happy to go with your approach instead.

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:61]: Ok, I'll go with number 1 for now then and keep URL-safe base64 encoding everywhere. At this point changing this behavior at a later point is really not difficult. Thanks for the feedback!

File: runtime-components/non-repudiation-api/src/test/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpointSpec.scala (hunk 1 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/test/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpointSpec.scala]: Addressed by https://github.com/digital-asset/daml/pull/8982/commits/cdd891222d7f92ba6a0da3aca5a3ab95a13f0045

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala (hunk 6 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/SignedPayloadsEndpoint.scala:72]: Addressed by https://github.com/digital-asset/daml/pull/8982/commits/9faaed6d13deedf47fd8f07b7ddd4dfce425dcb7

I'm not 100% sure that we should include the status code in this API, but we can review the decision at a later point.

Following the `{ "error": "..." }`/`{ "result": { ... } }` pattern actually makes sense, it will hopefully allow users to reuse some code, or at least to be familiar with the output format.

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 5 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:65]: Addressed by https://github.com/digital-asset/daml/pull/8982/commits/e7ef80b57672191ea6bc907dc5032d0d6b39271c

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 4 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:41]: Follow-up in https://github.com/digital-asset/daml/pull/8982#discussion_r585666749

File: runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:26]: This requires a few changes to the PostgreSQL part, I'll merge this as is and open a separate PR to guarantee that the back-end guarantees idempotency.

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala:26]: Tracked in https://github.com/digital-asset/daml/issues/9021
